### PR TITLE
Allow retry for 307 - temporary redirect - response code

### PR DESCRIPTION
**Pull request by andrejvanderzee**
We are running into problematic edge cases running vault-agent as sidecar container in production environments: When we provision vault or when vault has leader elections, the server returns a 307 - temporary redirect - back to the vault agent side container. As a result, the vault agent re-authenticates and fetches new secrets, instead of retrying to renew the lease of the vault token, and the old vault token is eventually revoked together with all its issues secrets. Unfortunately, some of our Kubernetes services do not have the ability to renew database sessions with new credentials, or respond too late, resulting in many authorization errors in our production environments. 

The problem is that `retryablehttp.DefaultRetryPolicy` function does not allow 307 to be retried. Consequently, the lifetime watcher for the vault token is _canceled_ and the vault agent re-authenticates and fetches new secrets. The default check-retry policy from `retryablehttp` makes sense for common use cases, but Vault is designed to return a 307 when hitting a standby node, and that should not deny a retry, like it does now. 

This PR wraps `retryablehttp.DefaultRetryPolicy` to add logic for allowing 307 retries. It only affects the vault agent but it might be useful to set the wrapped retry policy in `api.DefaultConfig()` and make it default behavior. With this PR, vault agent recovers from vault token lease renewals in case of a temporary 307:

```$ kubectl logs kibana-test-5fbc54df65-pd4gt vault-agent -f
==> Vault agent started! Log data will stream in below:
==> Vault agent configuration:
                     Cgo: disabled
               Log Level: trace
                 Version: Vault v1.6.0-dev
             Version Sha: 0adf0329808576601bafb2584b719e7e1c44545a+CHANGES
2021-02-11T10:22:50.157Z [INFO]  sink.file: creating file sink
2021-02-11T10:22:50.157Z [TRACE] sink.file: enter write_token: path=/home/vault/.vault-token
2021-02-11T10:22:50.157Z [TRACE] sink.file: exit write_token: path=/home/vault/.vault-token
2021-02-11T10:22:50.157Z [INFO]  sink.file: file sink configured: path=/home/vault/.vault-token mode=-rw-r-----
2021-02-11T10:22:50.158Z [INFO]  template.server: starting template server
2021-02-11T10:22:50.158Z [INFO]  template.server: no templates found
2021-02-11T10:22:50.158Z [INFO]  sink.server: starting sink server
2021-02-11T10:22:50.158Z [INFO]  auth.handler: starting auth handler
2021-02-11T10:22:50.158Z [INFO]  auth.handler: authenticating
2021-02-11T10:22:50.158Z [TRACE] auth.kubernetes: beginning authentication
2021-02-11T10:22:50.253Z [INFO]  auth.handler: authentication successful, sending token to sinks
2021-02-11T10:22:50.253Z [INFO]  auth.handler: starting renewal process
2021-02-11T10:22:50.253Z [TRACE] sink.file: enter write_token: path=/home/vault/.vault-token
2021-02-11T10:22:50.253Z [INFO]  sink.file: token written: path=/home/vault/.vault-token
2021-02-11T10:22:50.253Z [TRACE] sink.file: exit write_token: path=/home/vault/.vault-token
2021-02-11T10:22:50.264Z [INFO]  auth.handler: renewed auth token
2021-02-11T10:33:45.378Z [DEBUG] retry.check: retry for response code 503
2021-02-11T10:33:45.378Z [DEBUG] retry.backoff: min duration 1s, max duration 1.5s, attempt 0, next retry: 1.137832553s
2021-02-11T10:33:46.520Z [DEBUG] retry.check: retry for response code 503
2021-02-11T10:33:46.520Z [DEBUG] retry.backoff: min duration 1s, max duration 1.5s, attempt 1, next retry: 2.025932636s
2021-02-11T10:33:48.549Z [DEBUG] retry.check: retry for response code 307
2021-02-11T10:33:48.549Z [DEBUG] retry.backoff: min duration 1s, max duration 1.5s, attempt 2, next retry: 3.531723531s
2021-02-11T10:33:52.083Z [DEBUG] retry.check: retry for response code 503
2021-02-11T10:33:52.083Z [DEBUG] retry.backoff: min duration 1s, max duration 1.5s, attempt 3, next retry: 5.613898064s
2021-02-11T10:33:57.699Z [DEBUG] retry.check: retry for response code 503
2021-02-11T10:33:57.699Z [DEBUG] retry.backoff: min duration 1s, max duration 1.5s, attempt 4, next retry: 7.359382325s
2021-02-11T10:34:05.061Z [DEBUG] retry.check: retry for response code 307
2021-02-11T10:34:05.061Z [DEBUG] retry.backoff: min duration 1s, max duration 1.5s, attempt 5, next retry: 8.222880168s
2021-02-11T10:34:13.287Z [DEBUG] retry.check: retry for response code 503
2021-02-11T10:34:13.287Z [DEBUG] retry.backoff: min duration 1s, max duration 1.5s, attempt 6, next retry: 7.514124968s
2021-02-11T10:34:20.816Z [INFO]  auth.handler: renewed auth token
2021-02-11T10:45:15.921Z [INFO]  auth.handler: renewed auth token
2021-02-11T10:56:11.028Z [INFO]  auth.handler: renewed auth token
2021-02-11T11:07:06.136Z [INFO]  auth.handler: renewed auth token
2021-02-11T11:18:01.266Z [INFO]  auth.handler: renewed auth token
```

In the above logs, after 10 minutes the vault token renewal process starts. At that time, vault is sealed. After less than 1 minute the Vault is unsealed and the lease on the vault token is renewed successfully, leaving the already issued secrets unaffected.

### CODE REVIEW

**Comment by andrejvanderzee** [comment]
@jefferai  @HridoyRoy @ncabatoff 

Could you lease look at this? We are hitting this edge-case in production environments regularly!

**Comment by jefferai** [comment]
I don't quite follow what you mean by it not allowing 307 redirects to be retried. Do you mean it doesn't follow the redirects? Or it follows the redirect, this fails, then it doesn't retry?

Either way it seems like the right fix is in go-retryablehttp, not here.

**Comment by andrejvanderzee** [comment]
@jefferai The problem is that 307 are not retried at all. So when the vault server returns a 307, Vault clients do not retry and abort instead. This is a problem in context of vault agent as a side container, because it re-authenticates and revokes the old vault token and secrets.

**Comment by andrejvanderzee** [comment]
Filed a PR to `go-retryablehttp` for retrying 307 response codes:
https://github.com/hashicorp/go-retryablehttp/pull/123

**Comment by ncabatoff** [comment]
@andrejvanderzee can you elaborate on 

> the server returns a 307 - temporary redirect - back to the vault agent side container.

Given the preceding 503s and the scenario you describe, this sounds like it's a standby redirecting to the active node here: https://github.com/hashicorp/vault/blob/master/http/handler.go#L879

Though normally it would prefer to forward requests to the active node rather than resort to redirections.  Assuming you haven't explicitly disabled forwarding, there's probably an error in the logs of the vault server you're hitting that explains why forwarding failed.

The correct behaviour for a 307 is to follow the redirect, not retry.  So I don't think it's likely we're going to change this.  Instead, let's figure out why you're seeing the 307.

In addition to server logs, I'd be interested in seeing the contents of the redirect response.  You say it's pointing to the agent side container - can you show what that looks like?  I wouldn't expect the Vault servers to know the address of the agent.

**Review by arengoh — COMMENTED** [review]
Sorry .

**Comment by andrejvanderzee** [comment]
@ncabatoff We get a 307 indeed when the requests lands on a standby node. When this happens in the vault agent, the lifetime watcher abandons this for-loop with `return nil`:
https://github.com/hashicorp/vault/blob/master/api/lifetime_watcher.go#L337
As a result, the life watcher is _restarted_ and also the template manager (eventually revoking the old token and secrets).

In our setup, we have an ALB in front of Vault. When we provision vault or vault does a raft leader re-election, the ALB can hit a standby node and return a 307. As a result, the vault agent re-authenticates and the old vault token plus secrets are eventually revoked, causing problems to containers that rely on secrets to be valid util it terminates.

In other words, the vault agent cannot handle 307 without _completely_ restarting itself, causing problems in our production environments.

**Comment by ncabatoff** [comment]
But why is Agent even aware of the redirection?  If the standby returned a 307, the client used by agent should follow the redirection, it should hit the active node, and Agent should be none the wiser.  I guess the problem here is likely caused the ALB, and/or because you don't have access to the vault nodes without going through it.

But even before debugging that, why is Vault issuing the 307 in the first place?  The normal behaviour is for a standby to transparently forward the request the request to the active node using gRPC.  The client would only see a 307 if you've actively disabled forwarding, or if something else is going wrong.  Which is why I asked to see Vault server logs.

**Comment by andrejvanderzee** [comment]
The ALB just returns the 307 from the standby node that the ALB still thinks is healthy cause its health probe did not kick in yet. 

We would also prefer that the Vault server would internally redirect the call to the active node. Is something wrong with our configuration? This is our development setup:

```storage "raft" {
  path    = "/mnt/vault/storage/"
  node_id = "node_0"
  performance_multiplier = 5
  retry_join { leader_api_addr = "http://172.22.20.4:8200" }
  retry_join { leader_api_addr = "http://172.22.21.4:8200" }
  retry_join { leader_api_addr = "http://172.22.22.4:8200" }
}

disable_mlock = true

listener "tcp" {
  address = "172.22.20.4:8200"
  tls_disable = 1
}

listener "tcp" {
  address = "127.0.0.1:8200"
  tls_disable = 1
}

telemetry {
  prometheus_retention_time = "30s",
  disable_hostname = true
}

plugin_directory = "/etc/vault/vault-plugins"
api_addr = "http://172.22.20.4:8200"
cluster_addr = "https://172.22.20.4:8201"
```

**Comment by andrejvanderzee** [comment]
Hmmm, maybe we see this happening mostly when all Vault instances are in sealed state? What does Vault server return when it cannot internally redirect a call cause there is no active node?

**Comment by ncabatoff** [comment]
I believe a sealed Vault returns 503s, except for things like status endpoints.  Same for when there's no leader known by the node handling the request.

Normally forwarding will always be used unless a header disabling it is provided in the request, as per https://www.vaultproject.io/docs/concepts/ha#client-redirection.  

You might be able to work around your problem for now by setting api_addr to the LB addr, as discussed here: https://www.vaultproject.io/docs/concepts/ha#behind-load-balancers.  Though since you say the LB isn't aware of the standby node's state that might cause a redirect loop.

**Comment by andrejvanderzee** [comment]
@ncabatoff Indeed the problem arises when two out of three Vaults are sealed. These are the server logs of the only Vault instance that is active, while two others are sealed:

```
Feb 12 23:18:53 ip-172-22-22-4 vault[466164]: {"@level":"warn","@message":"Election timeout reached, restarting election","@module":"storage.raft","@timestamp":"2021-02-12T23:18:53.919157Z"}
Feb 12 23:18:53 ip-172-22-22-4 vault[466164]: {"@level":"info","@message":"entering candidate state","@module":"storage.raft","@timestamp":"2021-02-12T23:18:53.919238Z","node":{},"term":17033}
Feb 12 23:18:53 ip-172-22-22-4 vault[466164]: {"@level":"error","@message":"failed to make requestVote RPC","@module":"storage.raft","@timestamp":"2021-02-12T23:18:53.922096Z","error":"dial tcp 172.22.21.4:8201: connect: connection refused","target":{"Suffrage":0,"ID":"node_1","Address":"172.22.21.4:8201"}}
Feb 12 23:18:53 ip-172-22-22-4 vault[466164]: {"@level":"error","@message":"failed to make requestVote RPC","@module":"storage.raft","@timestamp":"2021-02-12T23:18:53.922369Z","error":"dial tcp 172.22.20.4:8201: connect: connection refused","target":{"Suffrage":0,"ID":"node_0","Address":"172.22.20.4:8201"}}
Feb 12 23:18:54 ip-172-22-22-4 vault[466164]: {"@level":"error","@message":"error during forwarded RPC request","@module":"core","@timestamp":"2021-02-12T23:18:54.983202Z","error":"rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing remote error: tls: internal error\""}
Feb 12 23:18:54 ip-172-22-22-4 vault[466164]: {"@level":"error","@message":"forward request error","@module":"core","@timestamp":"2021-02-12T23:18:54.983243Z","error":"error during forwarding RPC request"}
```

```
$ ./vault-test 
2021/02/13 00:18:54 redirect would cause protocol downgrade
```

The problem for us is that vault agent does not cope with the 307 and re-authenticates and fetches new secrets. Maybe it is our setup and I hope you can advice us howto fix this.

**Comment by andrejvanderzee** [comment]
@ncabatoff The reason for this PR actually started with our original problem in production environments: https://github.com/hashicorp/vault/issues/10918

And thanks for your time and looking into this!

**Comment by ncabatoff** [comment]
So the obvious question is why are your vault servers so unhealthy?  We now understand the reason for the 307s - the last unsealed node can't talk to anyone, and there's no quorum to elect a leader.  What caused the other nodes to seal themselves?

**Comment by andrejvanderzee** [comment]
@ncabatoff Indeed I got misled by the 307 when quorum cannot be reached by Vault servers. I understand now that Vault does request forwarding to standby nodes internally _unless_ there is no other healthy Vault instance. Still I do not fully understand why it returns a 307 redirect to the client in such a case, instead of some 5xx error code? 

But the actual problem I am chasing is still unresolved: The vault agent is not really resilient in it's vault token lease renewal process. For example, when we have a network problem between our k8s PODs and Vault that lasts for longer than about 1 minute, we see in the logs that the vault token lifetime watcher and the consul template manager restart and the "old" vault token and secrets are revoked (eventually). As a consequence, PODs end up with authorization problems (ideally PODs should be able to respond to renewals, but it is not always feasible to fetch the new secrets and restart database sessions without problems - depending on client SDK etc).

The VAULT_MAX_RETRIES solves this partially by having the request retried for a limited period. For us it would be very helpfull if the Vault agent would be able to retry to renew the Vault token's lease until it actually expires: Supposedly my vault agent manages a Vault token with a TTL of 6 days, then roughly after 4 days it would start the renewal process and still could be renewed for 2 days. But currently, after max retries is cancelled after roughly 1 minute (context deadline?), we get into the vault agent "restart"-situation and we end up with broken clients. The logic is here and should not be that tough to change:
https://github.com/hashicorp/vault/blob/master/api/lifetime_watcher.go#L337

**Comment by andrejvanderzee** [comment]
@ncabatoff Below you will find a trace of vault where I temporarily close a security group so that the agent cannot reach vault. As a result, we get the message in the trace-log: `auth.handler: lifetime watcher done channel triggered`. This means that vault agent just bails out and get's a new vault token instead of retrying (while the old token will expire and secrets are revoked).


```
$ kubectl logs kibana-test-999c58874-mn58b vault-agent -f
==> Vault agent started! Log data will stream in below:
==> Vault agent configuration:
                     Cgo: disabled
               Log Level: info
                 Version: Vault v1.6.1
             Version Sha: 6d2db3f033e02e70202bef9ec896360062b88b03
2021-02-15T12:36:41.634Z [INFO]  sink.file: creating file sink
2021-02-15T12:36:41.634Z [INFO]  sink.file: file sink configured: path=/home/vault/.vault-token mode=-rw-r-----
2021-02-15T12:36:41.635Z [INFO]  template.server: starting template server
2021-02-15T12:36:41.635Z [INFO]  auth.handler: starting auth handler
2021-02-15T12:36:41.635Z [INFO]  auth.handler: authenticating
2021/02/15 12:36:41.635525 [INFO] (runner) creating new runner (dry: false, once: false)
2021-02-15T12:36:41.635Z [INFO]  sink.server: starting sink server
2021/02/15 12:36:41.635988 [INFO] (runner) creating watcher
2021-02-15T12:36:41.787Z [INFO]  auth.handler: authentication successful, sending token to sinks
2021-02-15T12:36:41.787Z [INFO]  auth.handler: starting renewal process
2021-02-15T12:36:41.787Z [INFO]  template.server: template server received new token
2021/02/15 12:36:41.787580 [INFO] (runner) stopping
2021/02/15 12:36:41.787616 [INFO] (runner) creating new runner (dry: false, once: false)
2021/02/15 12:36:41.787764 [INFO] (runner) creating watcher
2021-02-15T12:36:41.787Z [INFO]  sink.file: token written: path=/home/vault/.vault-token
2021/02/15 12:36:41.787797 [INFO] (runner) starting
2021-02-15T12:36:41.798Z [INFO]  auth.handler: renewed auth token
2021/02/15 12:36:41.892062 [INFO] (runner) rendered "(dynamic)" => "/usr/share/kibana/ssl/elastic-ca.pem"
2021-02-15T12:38:48.345Z [INFO]  auth.handler: renewed auth token
2021-02-15T12:40:54.883Z [INFO]  auth.handler: renewed auth token
2021-02-15T12:43:01.427Z [INFO]  auth.handler: renewed auth token
2021-02-15T12:46:07.941Z [INFO]  auth.handler: lifetime watcher done channel triggered
2021-02-15T12:46:07.942Z [INFO]  auth.handler: authenticating
2021-02-15T12:46:15.119Z [INFO]  auth.handler: authentication successful, sending token to sinks
2021-02-15T12:46:15.119Z [INFO]  auth.handler: starting renewal process
2021-02-15T12:46:15.119Z [INFO]  template.server: template server received new token
2021/02/15 12:46:15.119654 [INFO] (runner) stopping
2021/02/15 12:46:15.119723 [INFO] (runner) creating new runner (dry: false, once: false)
2021-02-15T12:46:15.119Z [INFO]  sink.file: token written: path=/home/vault/.vault-token
2021/02/15 12:46:15.119758 [INFO] (runner) received finish
2021/02/15 12:46:15.119922 [INFO] (runner) creating watcher
2021/02/15 12:46:15.119972 [INFO] (runner) starting
2021-02-15T12:46:15.130Z [INFO]  auth.handler: renewed auth token
2021-02-15T12:48:24.929Z [INFO]  auth.handler: renewed auth token
2021-02-15T12:50:34.730Z [INFO]  auth.handler: renewed auth token
```

**Comment by andrejvanderzee** [comment]
@ncabatoff Below you will find a trace of vault where I temporarily close a security group so that the agent cannot reach vault. As a result, we get the message in the trace-log: auth.handler: lifetime watcher done channel triggered. This means that vault agent just bails out and get's a new vault token instead of retrying (while the old token will expire and secrets are revoked).

$ kubectl logs kibana-test-999c58874-mn58b vault-agent -f
==> Vault agent started! Log data will stream in below:
==> Vault agent configuration:
                     Cgo: disabled
               Log Level: info
                 Version: Vault v1.6.1
             Version Sha: 6d2db3f033e02e70202bef9ec896360062b88b03
2021-02-15T12:36:41.634Z [INFO]  sink.file: creating file sink
2021-02-15T12:36:41.634Z [INFO]  sink.file: file sink configured: path=/home/vault/.vault-token mode=-rw-r-----
2021-02-15T12:36:41.635Z [INFO]  template.server: starting template server
2021-02-15T12:36:41.635Z [INFO]  auth.handler: starting auth handler
2021-02-15T12:36:41.635Z [INFO]  auth.handler: authenticating
2021/02/15 12:36:41.635525 [INFO] (runner) creating new runner (dry: false, once: false)
2021-02-15T12:36:41.635Z [INFO]  sink.server: starting sink server
2021/02/15 12:36:41.635988 [INFO] (runner) creating watcher
2021-02-15T12:36:41.787Z [INFO]  auth.handler: authentication successful, sending token to sinks
2021-02-15T12:36:41.787Z [INFO]  auth.handler: starting renewal process
2021-02-15T12:36:41.787Z [INFO]  template.server: template server received new token
2021/02/15 12:36:41.787580 [INFO] (runner) stopping
2021/02/15 12:36:41.787616 [INFO] (runner) creating new runner (dry: false, once: false)
2021/02/15 12:36:41.787764 [INFO] (runner) creating watcher
2021-02-15T12:36:41.787Z [INFO]  sink.file: token written: path=/home/vault/.vault-token
2021/02/15 12:36:41.787797 [INFO] (runner) starting
2021-02-15T12:36:41.798Z [INFO]  auth.handler: renewed auth token
2021/02/15 12:36:41.892062 [INFO] (runner) rendered "(dynamic)" => "/usr/share/kibana/ssl/elastic-ca.pem"
2021-02-15T12:38:48.345Z [INFO]  auth.handler: renewed auth token
2021-02-15T12:40:54.883Z [INFO]  auth.handler: renewed auth token
2021-02-15T12:43:01.427Z [INFO]  auth.handler: renewed auth token
2021-02-15T12:46:07.941Z [INFO]  auth.handler: lifetime watcher done channel triggered
2021-02-15T12:46:07.942Z [INFO]  auth.handler: authenticating
2021-02-15T12:46:15.119Z [INFO]  auth.handler: authentication successful, sending token to sinks
2021-02-15T12:46:15.119Z [INFO]  auth.handler: starting renewal process
2021-02-15T12:46:15.119Z [INFO]  template.server: template server received new token
2021/02/15 12:46:15.119654 [INFO] (runner) stopping
2021/02/15 12:46:15.119723 [INFO] (runner) creating new runner (dry: false, once: false)
2021-02-15T12:46:15.119Z [INFO]  sink.file: token written: path=/home/vault/.vault-token
2021/02/15 12:46:15.119758 [INFO] (runner) received finish
2021/02/15 12:46:15.119922 [INFO] (runner) creating watcher
2021/02/15 12:46:15.119972 [INFO] (runner) starting
2021-02-15T12:46:15.130Z [INFO]  auth.handler: renewed auth token
2021-02-15T12:48:24.929Z [INFO]  auth.handler: renewed auth token
2021-02-15T12:50:34.730Z [INFO]  auth.handler: renewed auth token

**Comment by ncabatoff** [comment]
I'm going to close this because you've dropped the original request and your most recent updates really belong in #10918.
